protected/humhub/exceptions/InvalidTypeExceptionTrait.php

```diff
@@ -13,56 +13,23 @@
  */
 trait InvalidTypeExceptionTrait
 {
-    //  public properties
-
-    public string $methodName;
-    public $parameter;
-    public array $validType = [];
-
-    /**
-     * @var mixed|null
-     */
-    public $givenValue;
-
-    /**
-     * @param string $method
-     * @param int|array $parameter = [
-     *                                     int => string,       // position, or [ position => name ] of the  argument
-     *                                     ]
-     * @param array|string|null $validType
-     * @param null $givenValue
-     */
-    public function __construct(
-        $method = '',
-        $parameter = null,
-        $validType = [],
-        $givenValue = null,
-        $nullable = false,
-        $code = 0,
-        $previous = null
-    ) {
-
-        $this->methodName = $method;
-        $this->parameter = $parameter;
-        $this->validType = (array)($validType ?? ['mixed']);
-        $this->givenValue = $givenValue;
+    use InvalidArgumentExceptionTrait;
```

Comment:
@martin-rueegg I also see in your other PR that you use traits very intensively. So far, we have done this only in rare cases.

Is it really necessary to this extent? Where are the advantages here? I'm a bit worried that it complicates things again if it's spread over multiple class/traits.

Reply:
Traits have the following advantages
- Reduce code redundancy
- Allow classes of different inheritance to behave exactly the same, e.g. when implementing an interface
- have been introduced to PHP v5.4.0 and as such should be known by a common PHP developer and are also widely supported by IDEs

In contrast, **Yii2 behaviors** seem much more complex to me and 
  - are only supported by some IDEs [with plugin]) and
  - require more configuration and
  - more memory, and
  - are only available to components
  - The advantage of behaviors is that
    -  they can be sneaked into common objects by altering the container's definition and as such
    - can be provided by extensions (traits could be provided by extensions, too, though).
    - Also, they allow to load code only if really required, which for me is the biggest advantage, unless the behavior is called on nearly every invocation and every object it implements it. For example, the polymorphic relation is used extensively and should, IMHO, better be used as a trait which would reduce memory and function calls.

The main disadvantage of traits for me are
- no support for constants before php v8.2
- defined variables must match type and initial value of any implementing or extending class
- you cannot just fallback on `parent::method()` but need to use either two different method names or manually rename the trait's method on import. However, for behaviors, there are also caveats in this scenario. Renaming on import is not a big deal. What annoys me most here, is, that if you ctrl-click on the alias, it takes you to the import statement and there you have to ctrl-click again on the original method name. And going back takes the two steps as well. 

Now, whether or not they are necessary? No, absolutely not. No trait can do what you could not do with copy-and-past (and potentially implementing interfaces to test the instance for a given capability).

Of course, the disadvantages of code-duplication is well known: more maintenance and eventually divergence.

What I find really cool with traits though, is when defining an Interface, e.g. `ArrayAccessInterface`, to also create a trait alongside it, i.e. `ArrayAccessTrait`. Now a class can implement the interface by adding 1.5 lines of code:

- `implements ArrayAccessInterface`
- `use ArrayAccessTrait`

Done.

The proposed StateService implementation uses this to implement Editable, Deletable, Draftable and other states. See https://github.com/humhub/humhub/pull/6430/files#diff-dd4d14a7d1bda6fa1dc4821d285fcbb1e336ee7dccd43951b8a6443b43569775R1

If this would have to be maintained in all the tables that will use the state field, this would be a lot of duplicate code to maintain. Also, if a new test for one state is required, this can be added to the trait and you're set.

By the way, even Yii2 itself is using exactly this design pattern in it's key components and on different levels (see `Query` and `ActiveQuery`):
- `\yii\base\Model`
  ```php 
  class Model extends Component implements StaticInstanceInterface, IteratorAggregate, ArrayAccess, Arrayable
  {
    use ArrayableTrait;
    use StaticInstanceTrait;
  ```
- `\yii\db\ActiveQuery`
  ```php 
  class ActiveQuery extends Query implements ActiveQueryInterface
  {
    use ActiveQueryTrait;
    use ActiveRelationTrait;
  ```
- `\yii\db\Query`
  ```php 
  class Query extends Component implements QueryInterface, ExpressionInterface
  {
    use QueryTrait;
  ```

Now for this current example of the Exception, I don't think an Interface is required.

Happy to discuss further. But eventually it's up to you to decide, if traits are eligible or in which particular cases.

Reply:
You're welcome!

If I may add: I would assume that if Interfaces (and potentially Traits) would have been used more often in the code, the many different implementations of the `can*` methods would have been a bit more standardized. Some models use `canRead()`, others use `canView()`, although it's essentially the same. Some allow `User` as first parameter, others only `user_id`or vice versa.

The `FindInstanceInterface` (#6503) together with the `EditableInterface` and friends (#6451) are an attempt to streamline this.